rp2040-hal/examples/i2c_async_irq.rs

```diff
@@ -0,0 +1,143 @@
+//! # I²C Example
+//!
+//! This application demonstrates how to talk to I²C devices with an RP2040.
+//! in an Async environment.
+//!
+//! This example sets a `waker_setter` and is able to sleep when the task is pending on i2c
+//! updates..
+//!
+//! It may need to be adapted to your particular board layout and/or pin assignment.
+//!
+//! See the `Cargo.toml` file for Copyright and license details.
+
+#![no_std]
+#![no_main]
+
+use core::{cell::RefCell, task::Waker};
+
+use critical_section::Mutex;
+
+// Ensure we halt the program on panic (if we don't mention this crate it won't
+// be linked)
+use panic_halt as _;
+
+// Alias for our HAL crate
+use rp2040_hal as hal;
+
+// Some traits we need
+use hal::fugit::RateExtU32;
+
+// Import required types & traits.
+use eh_async_1_0_alpha::i2c::I2c;
+use hal::{
+    gpio::{FunctionI2C, Pin, PullUp},
+    pac::{self, interrupt},
+    Clock,
+};
+
+/// The linker will place this boot block at the start of our program image. We
+/// need this to help the ROM bootloader get our code up and running.
+/// Note: This boot block is not necessary when using a rp-hal based BSP
+/// as the BSPs already perform this step.
+#[link_section = ".boot2"]
+#[used]
+pub static BOOT2: [u8; 256] = rp2040_boot2::BOOT_LOADER_GENERIC_03H;
+
+/// External high-speed crystal on the Raspberry Pi Pico board is 12 MHz. Adjust
+/// if your board has a different frequency
+const XTAL_FREQ_HZ: u32 = 12_000_000u32;
+
+/// Mutex holding an optional waker for the i2c peripiheral.
+static WAKER: Mutex<RefCell<Option<Waker>>> = Mutex::new(RefCell::new(None));
+
+/// Waker setter method updating the waker of the async task.
+pub fn i2c_waker_setter(waker: Waker) {
+    critical_section::with(|cs| {
+        *WAKER.borrow_ref_mut(cs) = Some(waker);
+    });
+}
+
+#[interrupt]
+fn I2C1_IRQ() {
+    let i2c1 = unsafe { &rp2040_hal::pac::Peripherals::steal().I2C1 };
```

Comment:
Should add a note here about why this is acceptable. Something about no other code (main loop or interrupt) will use I2C1 concurrently, and thus you won't get a race when modifying the ic_intr_mask register on I2C1.

Reply:
One things is the clearing of the flags from the IRQ, another one is to soundly manage the waker.

I can't think of a solution that does not invole stealing for the former since we give that instance to the driver.

For the latter, I beleive this is executor specific, so we cannot really have a one size fits all. At worst, there should exist some `critical-section` based channel implementations. I haven't checked though.

Reply:
I don't think the waker stuff is necessarily executor specific. We could for example use the utility classes from embassy_sync that seem to be executor independent:

> Futures from this crate can run on any executor.

(From https://docs.embassy.dev/embassy-sync/git/default/index.html)

I must admit that I don't fully understand the implications, and that I still need to learn a lot about async rust. But having all the low level stuff shown in this example in user code doesn't seem very attractive.
Look at the corresponding embassy code: https://github.com/embassy-rs/embassy/blob/main/examples/rp/src/bin/i2c_async.rs
Not as single `unsafe`, no mention of a waker. Sure, that's embassy specific code. But I see no reason why it shouldn't be portable to other executors. In fact, as I understand it, embassy is designed so the HALs it provides should actually be portable to other executors.

Reply:
> We could for example use the utility classes from embassy_sync that seem to be executor independent

Potentially, `embassy-sync` defines an [`AtomicWaker`](https://github.com/embassy-rs/embassy/blob/main/embassy-sync/src/waitqueue/atomic_waker.rs#L8) type that essentially does what this example does without embassy. I'm not convinced adding this dependency would solve our issues though.

> Not as single unsafe, no mention of a waker. Sure, that's embassy specific code

It's hidden behind a macro but there definitely is unsafe and wakers involved. The `bind_interruts` macro generates an unsafe IRQ handlers that calls an unsafe method from the driver.

I could use this approach too, at least with the rp2040-hal owned static storage for the storage. That'd save use from needing to expose those callbacks.

I'll explore that. A thing I'd like to avoid is having an unused I2C instance from taking space for its static atomic that'd never be used.

https://github.com/embassy-rs/embassy/blob/main/embassy-rp/src/lib.rs#L105-L124
https://github.com/embassy-rs/embassy/blob/main/embassy-rp/src/i2c.rs#L335-L341

Reply:
> It's hidden behind a macro but there definitely is unsafe and wakers involved.

Sure - if you dig deep enough, everything is using unsafe. The point is: The HAL does the heavy lifting, the user can use a safe abstraction.

> A thing I'd like to avoid is having an unused I2C instance from taking space for its static atomic that'd never be used.

Keeping the abstractions zero-cost, especially when they are not used, is a worthy goal. That said, I wouldn't object to a few tens of bytes constant RAM overhead if it significantly improves usability.

Reply:
A little update.
I made progress and the implementation looks better than I anticipated.
However, as I progressed on it I noticed several bugs so I'm solving them as they are discovered.
Mostly around input validation, error handling and stop generation and processing.